src/main/scala/com.snowplowanalytics.iglu/ctl/IgluctlConfig.scala

```diff
@@ -20,19 +25,16 @@ case class IgluctlConfig(description: Option[String],
 
 object IgluctlConfig {
   /** Trait common to S3cp and Push commands */
-  sealed trait IgluctlAction {
-    def process: Result
-  }
+  sealed trait IgluctlAction
 
-  object IgluctlAction {
-    case class S3Cp(command: Command.StaticS3Cp) extends IgluctlAction {
-      def process: Result =
-        commands.S3cp.process(command.input, command.bucket, command.s3Path, command.accessKeyId, command.secretKeyId, command.profile, command.region, command.skipSchemaLists)
-    }
-
-    case class Push(command: Command.StaticPush) extends IgluctlAction {
-      def process: Result =
-        commands.Push.process(command.input, command.registryRoot, command.apikey, command.public, command.legacy)
+  def process(httpClient: Client[IO], config: IgluctlAction): Result = {
+    config match {
+      case S3Cp(command) => commands.S3cp.process(command.input, command.bucket, command.s3Path, command.accessKeyId, command.secretKeyId, command.profile, command.region, command.skipSchemaLists)
+      case Push(command) => commands.Push.process(command.input, command.registryRoot, command.apikey, command.public, command.legacy, httpClient)
```

Comment:
Currently, in `Main`, we build an http client and then differentiate between commands that can be run directly and commands that need to be run using the `withClient` function.

I think this works, but I had a different idea.

Seeing how the client is command-specific, what if it is created as part of the `Command` class itself? Then, all the different `command.???.process` functions can be simplified to accept a single `command` argument, which will already have everything they need. And we'll not need to pattern match here.

I haven't properly evaluated all the implications of this. But on the surface it seems like it's tidier.